File: Strings/valid-palindrome/valid-palindrome.go
// https://leetcode.com/problems/valid-palindrome/

package main

import "strings"

func isPalindrome(s string) bool {
	s = strings.ToLower(s)
	b := []byte(s)
	i := 0
	j := len(b) - 1

	for i < j {
		if i < j && !isAlphanumeric(b[i]) {
			i++
		} else if i < j && !isAlphanumeric(b[j]) {
			j--
		} else if b[i] != b[j] {
			return false
		} else {
			i++
			j--
		}
	}

	return true
}

func isAlphanumeric(b byte) bool {
	return (b >= 'a' && b <= 'z') || (b >= '0' && b <= '9')
}

File: Strings/valid-palindrome/valid-palindrome.rs

// https://leetcode.com/problems/valid-palindrome/

struct Solution;

impl Solution {
  pub fn is_palindrome(s: String) -> bool {
    let lower_case = s.chars().filter(|c| c.is_alphanumeric()).collect::<String>().to_lowercase();
    let chars: Vec<char> = lower_case.chars().collect();

    let half = lower_case.len() / 2;

    for i in 0..half {
      if chars[i] != chars[chars.len() - i - 1] {
        return false;
      }
    }

    true
  }
}

fn main() {
  let s = String::from("A man, a plan, a canal: Panama");
  let result = Solution::is_palindrome(s);
  println!("{}", result);
}
